### DISCUSSION TITLE
Please help with query.where

### DISCUSSION THREAD

**Discussion by PeterCodesSoftware** [GodotNuts/GodotFirebase#432] [Q&A] [answered]
Godot Version 4.2.2

Note: I'm not good at reading documentation yet so i might have made an obvious error

I'm using Firestroe to store different "games" that are scouted. I want to be able to get all of the documents that were scouted on the same team. My query doesn't work when I add:
`query.where("team_num", FirestoreQuery.OPERATOR.EQUAL, 1)`
It gives me the error "Trying to return a value of type Nil from a function which the return type is Array" When i remove this line, my code works (except for the fact that I get all the data, not just one team's data)
This is where Godot says there was a problem (in line 128 of firestore.gd):
`func query(query : FirestoreQuery) -> Array:
	var task : FirestoreTask = FirestoreTask.new()
	task.action = FirestoreTask.Task.TASK_QUERY
	var body : Dictionary = { structuredQuery = query.query }
	var url : String = _base_url + _extended_url + _query_suffix

	task.data = query
	task._fields = JSON.stringify(body)
	task._url = url
	_pooled_request(task)
	return await _handle_task_finished(task) # this is the line that has the error "Trying to return a value of type Nil from a function which the return type is Array" attached to it`

Below is the full code of my script that does the querying and pictures of my firebase are also attached

<img width="806" alt="Screenshot 2024-09-15 at 12 44 54 PM" src="https://github.com/user-attachments/assets/c062e763-d960-4d2e-b859-4ce0dfc0e149">
<img width="636" alt="Screenshot 2024-09-15 at 12 44 59 PM" src="https://github.com/user-attachments/assets/e8525f31-99fd-4264-8de1-29eae6f7af2b">

`
extends Control

@onready var team_num_input: LineEdit = $PanelContainer/MarginContainer/VBoxContainer/TeamNum
@onready var comptetition_input: OptionButton = $PanelContainer/MarginContainer/VBoxContainer/Competition
@onready var data_display: RichTextLabel = $PanelContainer/MarginContainer/VBoxContainer/DataDisplay

var team_num
var competition

func _ready():
	var defaults: FirestoreCollection = await Firebase.Firestore.collection("defaults")
	var defaults_doc: FirestoreDocument = await defaults.get_doc("defaults_doc")
	var comp_list = defaults_doc.get_value("comp_list")["values"]

	for i in range(comp_list.size()):
		comptetition_input.add_item(comp_list[i]["stringValue"], i + 1)

func _on_back_pressed():
	get_tree().change_scene_to_file("res://ui/main_menu/main_menu.tscn")

func _on_search_pressed():
	team_num = team_num_input.text
	competition = comptetition_input.text
	
	if check_false_search():
		return
	
	var query: FirestoreQuery = FirestoreQuery.new()
	query.from(competition)
	query.where("team_num", FirestoreQuery.OPERATOR.EQUAL, team_num.to_int())
	query.limit(20)
	
	print("querying")
	var match_datas = await Firebase.Firestore.query(query) # this line crashes when i have query.where
	print("querying over")
	var avg_speaker = 0
	
	for data : FirestoreDocument in match_datas:
		print(data)
		var speaker_value = data.get_value("auto")["values"][0]["integerValue"].to_int()
		print(speaker_value)
		avg_speaker += speaker_value
	
	var num_matches = match_datas.size()
	
	avg_speaker /= num_matches
	
	data_display.text = avg_speaker

func check_false_search():
	var team_num_error = false
	var competition_error = false
	var error_messages = []
	
	if not (team_num.is_valid_int() and (0 < team_num.to_int()) and (team_num.to_int() < 10000)):
		team_num_error = true
		error_messages.append("Please enter a valid team number")
		team_num_input.text = ""
	
	if comptetition_input.selected == 0:
		competition_error = true
		error_messages.append("Please select a competition")
	
	if team_num_error or competition_error:
		var local_error_popup = Global.popup.instantiate()
		local_error_popup.init(error_messages)
		get_tree().root.add_child(local_error_popup)
		return true
	
	return false

`

**Comment by WolfgangSenff** [comment]
There must be a bug somewhere in there. I'll see if I can look into it at some point maybe next week, but I dunno.

**Reply by PeterCodesSoftware** [reply]
Ok, I would really appreciate it if you could do that, as this is a project for my FIRST Robotics team. Just know that even if it takes over a month to look at it, it will still be appreciated.

**Reply by WolfgangSenff** [reply]
Actually, I think we might have someone who has already fixed it. They didn't get my message on here about the bug though, let me see if I can merge it.

**Reply by PeterCodesSoftware** [reply]
Awesome, how long do you think that will take?

**Reply by WolfgangSenff** [reply]
I think I'm wrong actually. I'll have to look into it. I'm still not sure how long it'll take. I did notice you have some code that's out of date though, and I believe a fix has already been made for it. You should get latest from github and try it again - you should never have to deal with the integerValues or anything like that. All things that come out of `get_value` should be the exact format you put them into it.

**Reply by PeterCodesSoftware** [reply]
I updated to 6.0.2, and now getting the files from the query is much easier. But, it still crashes with .where in it, same error message and location.

**Reply by PeterCodesSoftware** [reply]
Im just checking in, wondering if you still plan to fix this issue soon.

**Reply by WolfgangSenff** [reply, marked as answer]
I haven't really had a chance to look at it. However, I did realize it might just be an issue with your setup of Firebase itself (in the Firebase console). Here is something to try to make this work for yourself, without having to update your code anymore than you already have (per the above):

1. Go into the Firebase console on the web and go to Firestore. Then, go to the Indexes tab at the top. Next, click on single field at the top right of that page.
2. Click add exemption, and then fill out the fields so it looks roughly like this, except with defaults as the collection ID, and team_num as the field path. Make sure all the values are set to true for each of the ones that you can see in my pic.
<img width="976" alt="Screenshot 2024-10-08 at 10 13 54 AM" src="https://github.com/user-attachments/assets/3ea16bd5-f03c-4345-9943-5f923760d29b">
3. Try it out again and see if that makes it start working. If not, I guess it's possible you'd need an index to be created in the composite tab, but I don't think so, based on what you showed.

Let me know if that works.

**Comment by PeterCodesSoftware** [comment]
It worked! Thank you so much, this means so much to me because this was a big project for a school club. And while I don't think you need to know this, it was not the "default" collection that was the one that holds my data, it is my "test2024" collection and that change made it work.

**Reply by WolfgangSenff** [reply]
Great! I'm honestly still not sure why this is needed, and it bothers me that it is. It also came up recently on the Discord server for other reasons, so at least I had it in mind when seeing your previous follow-up!
